File: Vibe MessagesExtension/Models/NewsItem.swift
//
//  NewsItem.swift
//  Vibe MessagesExtension
//
//  Created on 1/29/26.
//

import Foundation

struct NewsItem: Identifiable, Codable {
    let id: String
    let headline: String
    let imageUrl: String?
    let source: String
    let url: String
    let publishedAt: Date
    let vibeScore: Int
    let batch: String // "morning", "noon", "evening"
    let createdAt: Date
    
    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case headline
        case imageUrl
        case source
        case url
        case publishedAt
        case vibeScore
        case batch
        case createdAt
    }
    
    var isJustIn: Bool {
        // Consider "Just In" if created within the last 2 hours
        Date().timeIntervalSince(createdAt) < 7200
    }
    
    var timeAgo: String {
        let interval = Date().timeIntervalSince(publishedAt)
        let hours = Int(interval / 3600)
        
        if hours < 1 {
            let minutes = Int(interval / 60)
            return "\(minutes)m ago"
        } else if hours < 24 {
            return "\(hours)h ago"
        } else {
            let days = Int(interval / 86400)
            return "\(days)d ago"
        }
    }
}

File: backend/src/models/NewsItem.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface INewsItem extends Document {
    headline: string;
    imageUrl?: string;
    source: string;
    url: string;
    publishedAt: Date;
    vibeScore: number;
    batch: 'morning' | 'noon' | 'evening';
    createdAt: Date;
}

const newsItemSchema = new Schema<INewsItem>(
    {
        headline: { type: String, required: true },
        imageUrl: { type: String },
        source: { type: String, required: true },
        url: { type: String, required: true },
        publishedAt: { type: Date, required: true },
        vibeScore: { type: Number, default: 0 },
        batch: {
            type: String,
            enum: ['morning', 'noon', 'evening'],
            required: true
        },
        createdAt: { type: Date, default: Date.now, expires: '24h' } // Auto-delete after 24h
    },
    { timestamps: true }
);

// Index for efficient sorting
newsItemSchema.index({ createdAt: -1 });

export const NewsItem = mongoose.model<INewsItem>('NewsItem', newsItemSchema);
